### ISSUE TITLE
(1453127) FreeBSD: crash of single-process Waterfox, mozilla::dom::HTMLMediaElement::ShutdownDecoder() () from /usr/local/lib/waterfox/libxul.so

### ISSUE THREAD

**Issue by grahamperrin** [BrowserWorks/waterfox#767]
- Waterfox 56.2.3 on FreeBSD-CURRENT
- experimental (proof of concept) [up-to-eleven](https://github.com/padenot/up-to-eleven#user-content-up-to-eleven) 0.2 by @padenot

### Steps

1. Switch Waterfox to single-process mode
2. accept the invitation to restart
3. quit, to avoid #766 
4. again, start
5. install the extension
6. about:memory
7. close all other tabs
8. new tab
9. https://www.youtube.com/watch?v=23lRkdDXqY0
10. during playback, click the up-to-eleven icon
11. bring the about:memory tab to front
12. tab context menu
13. close tabs to the right
14. minimise memory usage
15. observe the confirmation

### Expected

16. continued run of the application

### Actual result

16. crash. 

### Reproducibility

Firefox 62.0.3 (64-bit) in single-process mode on the same system: 

- not reproducible. 

Firefox ESR 60.x: 

- not yet yested. 

Waterfox 56.2.3 on Kubuntu: 

- not yet reproducible with exactly the same steps, but it _was_ quite easy to crash the application very soon after first use of the extension.

**Comment by grahamperrin** [comment]
Not reproducible with: 

- Waterfox 56.2.3 on Window 7
- outdated Firefox ESR 52.1.0 on Windows 7
- Waterfox 56.2.2, 56.2.3 and 56.2.4 on Mac OS X 10.9.5
- outdated Firefox ESR 52.9.0 on Mac OS X 10.9.5
- Firefox ESR 60.2.2 on Mac OS X 10.9.5. 

Back to the Kubuntu case: 

> not yet reproducible with exactly the same steps, but it was quite easy to crash the application very soon after first use of the extension.

If we can't make the issue more (or consistently) reproducible with 56.2.3 or 56.2.4 on Linux, I'll close this as _wontfix_.

**Comment by grahamperrin** [comment]
Reproducible with Waterfox 56.2.5 on FreeBSD-CURRENT. (Patched, incidentally, for Mozilla bug [1447867 - Use shm_open for IPC shared memory on POSIX platforms](https://bugzilla.mozilla.org/show_bug.cgi?id=1447867).) 

Single-process, safe mode, about:debugging temporary load of the extension. As the crash does not occur until some time _after_ use of the extension, I imagine that nothing useful can be gained from this approach to debugging. The crash: 

```text
$ date ; uname -v
Thu 22 Nov 2018 04:41:08 GMT
FreeBSD 13.0-CURRENT r340689 GENERIC 
$ pkg query %v %R waterfox
56.2.5
$ waterfox --safe-mode -p up-to-eleven
Fontconfig warning: "/usr/local/etc/fonts/local.conf", line 1093: saw number, expected matrix
Fontconfig warning: "/usr/local/etc/fonts/local.conf", line 1093: saw number, expected matrix
1542861707209   addons.xpi      WARN    Addon with ID {242c2018-6cd5-4c97-ac19-3f248ccb8272} already installed, older version will be disabled
1542861707925   addons.webextension.{242c2018-6cd5-4c97-ac19-3f248ccb8272}      WARN    Please specify whether you want browser_style or not in your browser_action options.
Fontconfig warning: "/usr/local/etc/fonts/local.conf", line 1093: saw number, expected matrix
JavaScript error: resource://gre/modules/XPCOMUtils.jsm, line 192: TypeError: can't redefine non-configurable property "Services"
Segmentation fault (core dumped)
$ 
```

----

Not reproducible with 56.2.5 on Kubuntu, so I'm closing this issue. 

@padenot if you can guess a possible cause I'll be very grateful. tl;dr in single-process mode the browser crashes after closing a (YouTube) tab where the extension was used. It's a lot to ask because sadly :-1: about:crashes was removed from Waterfox. 

I set aside the `waterfox.core` dump file from the crash above, might take a look later although to be honest I'm stumbling in the dark with gdb and the like.

**Comment by padenot** [comment]
I don't see anything on the top of my head. However, if you can repro in `gdb`, `thread apply all bt` should output lots of stacks, and that would help.

**Comment by grahamperrin** [comment]
@padenot thanks :+1: 

@jbeich FYI ([not expecting support](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=226680#c3) with the deleted port) tl;dr: 

- a reproducible, _extraordinary_ crash 
- an extension that's experimental, but of great interest to me 
- **not** reproducible with Firefox on -CURRENT. 

### Environment

```text
$ date ; uname -v
Sun 25 Nov 2018 11:09:00 GMT
FreeBSD 13.0-CURRENT r340689 GENERIC 
$ poudriere jail -i -j current | grep -i version
Jail version:      13.0-CURRENT 1300003
Jail vcs version:  r340689
$ pkg query '%o %v %R' waterfox firefox
www/waterfox 56.2.5 poudriere
www/firefox 63.0.3_3,1 poudriere
$ 
```

### Crash

```text
Thread 1 received signal SIGSEGV, Segmentation fault.
0x0000000806be3193 in mozilla::dom::HTMLMediaElement::ShutdownDecoder() () from /usr/local/lib/waterfox/libxul.so
```
…

```text
(gdb) bt
#0  0x0000000806be3193 in mozilla::dom::HTMLMediaElement::ShutdownDecoder() () from /usr/local/lib/waterfox/libxul.so
#1  0x0000000806bee373 in mozilla::dom::HTMLMediaElement::~HTMLMediaElement() () from /usr/local/lib/waterfox/libxul.so
#2  0x0000000806c28250 in mozilla::dom::HTMLVideoElement::~HTMLVideoElement() () from /usr/local/lib/waterfox/libxul.so
#3  0x000000080528c03c in SnowWhiteKiller::~SnowWhiteKiller() () from /usr/local/lib/waterfox/libxul.so
#4  0x000000080528eb59 in nsCycleCollector::BeginCollection(ccType, nsICycleCollectorListener*) () from /usr/local/lib/waterfox/libxul.so
#5  0x000000080528e4e7 in nsCycleCollector::Collect(ccType, js::SliceBudget&, nsICycleCollectorListener*, bool) () from /usr/local/lib/waterfox/libxul.so
#6  0x000000080528f980 in nsCycleCollector_collect(nsICycleCollectorListener*) () from /usr/local/lib/waterfox/libxul.so
#7  0x000000080617502b in nsJSEnvironmentObserver::Observe(nsISupports*, char const*, char16_t const*) () from /usr/local/lib/waterfox/libxul.so
#8  0x00000008052acf8f in nsObserverList::NotifyObservers(nsISupports*, char const*, char16_t const*) () from /usr/local/lib/waterfox/libxul.so
#9  0x00000008052aea51 in nsObserverService::NotifyObservers(nsISupports*, char const*, char16_t const*) () from /usr/local/lib/waterfox/libxul.so
#10 0x00000008052a090b in (anonymous namespace)::MinimizeMemoryUsageRunnable::Run() () from /usr/local/lib/waterfox/libxul.so
#11 0x00000008052fc9bd in nsThread::ProcessNextEvent(bool, bool*) () from /usr/local/lib/waterfox/libxul.so
#12 0x00000008052fe4b7 in NS_ProcessNextEvent(nsIThread*, bool) () from /usr/local/lib/waterfox/libxul.so
#13 0x0000000805658226 in mozilla::ipc::MessagePump::Run(base::MessagePump::Delegate*) () from /usr/local/lib/waterfox/libxul.so
#14 0x00000008056205ba in MessageLoop::Run() () from /usr/local/lib/waterfox/libxul.so
#15 0x00000008073405a6 in nsBaseAppShell::Run() () from /usr/local/lib/waterfox/libxul.so
#16 0x00000008083c4799 in nsAppStartup::Run() () from /usr/local/lib/waterfox/libxul.so
#17 0x000000080845b58d in XREMain::XRE_mainRun() () from /usr/local/lib/waterfox/libxul.so
#18 0x000000080845bdcd in XREMain::XRE_main(int, char**, mozilla::BootstrapConfig const&) () from /usr/local/lib/waterfox/libxul.so
#19 0x000000080845c2e4 in XRE_main(int, char**, mozilla::BootstrapConfig const&) () from /usr/local/lib/waterfox/libxul.so
#20 0x000000000102c6e9 in main ()
(gdb) c
Continuing.

Thread 1 received signal SIGSEGV, Segmentation fault.
thr_kill () at thr_kill.S:3
3       RSYSCALL(thr_kill)
(gdb) bt
#0  thr_kill () at thr_kill.S:3
#1  0x00000008013c5e54 in __raise (s=11) at /usr/src/lib/libc/gen/raise.c:52
#2  0x0000000808450874 in nsProfileLock::FatalSignalHandler(int, __siginfo*, void*) () from /usr/local/lib/waterfox/libxul.so
#3  0x0000000808d53bdc in void WasmFaultHandler<(Signal)0>(int, __siginfo*, void*) () from /usr/local/lib/waterfox/libxul.so
#4  0x00000008011e44c4 in handle_signal (actp=0x7fffffffd040, sig=11, info=0x7fffffffd430, ucp=0x7fffffffd0c0) at /usr/src/lib/libthr/thread/thr_sig.c:246
#5  0x00000008011e3a22 in thr_sighandler (sig=11, info=0x7fffffffd430, _ucp=0x7fffffffd0c0) at /usr/src/lib/libthr/thread/thr_sig.c:191
#6  <signal handler called>
#7  0x0000000806be3193 in mozilla::dom::HTMLMediaElement::ShutdownDecoder() () from /usr/local/lib/waterfox/libxul.so
#8  0x0000000806bee373 in mozilla::dom::HTMLMediaElement::~HTMLMediaElement() () from /usr/local/lib/waterfox/libxul.so
#9  0x0000000806c28250 in mozilla::dom::HTMLVideoElement::~HTMLVideoElement() () from /usr/local/lib/waterfox/libxul.so
#10 0x000000080528c03c in SnowWhiteKiller::~SnowWhiteKiller() () from /usr/local/lib/waterfox/libxul.so
#11 0x000000080528eb59 in nsCycleCollector::BeginCollection(ccType, nsICycleCollectorListener*) () from /usr/local/lib/waterfox/libxul.so
#12 0x000000080528e4e7 in nsCycleCollector::Collect(ccType, js::SliceBudget&, nsICycleCollectorListener*, bool) () from /usr/local/lib/waterfox/libxul.so
#13 0x000000080528f980 in nsCycleCollector_collect(nsICycleCollectorListener*) () from /usr/local/lib/waterfox/libxul.so
#14 0x000000080617502b in nsJSEnvironmentObserver::Observe(nsISupports*, char const*, char16_t const*) () from /usr/local/lib/waterfox/libxul.so
#15 0x00000008052acf8f in nsObserverList::NotifyObservers(nsISupports*, char const*, char16_t const*) () from /usr/local/lib/waterfox/libxul.so
#16 0x00000008052aea51 in nsObserverService::NotifyObservers(nsISupports*, char const*, char16_t const*) () from /usr/local/lib/waterfox/libxul.so
#17 0x00000008052a090b in (anonymous namespace)::MinimizeMemoryUsageRunnable::Run() () from /usr/local/lib/waterfox/libxul.so
#18 0x00000008052fc9bd in nsThread::ProcessNextEvent(bool, bool*) () from /usr/local/lib/waterfox/libxul.so
#19 0x00000008052fe4b7 in NS_ProcessNextEvent(nsIThread*, bool) () from /usr/local/lib/waterfox/libxul.so
#20 0x0000000805658226 in mozilla::ipc::MessagePump::Run(base::MessagePump::Delegate*) () from /usr/local/lib/waterfox/libxul.so
#21 0x00000008056205ba in MessageLoop::Run() () from /usr/local/lib/waterfox/libxul.so
#22 0x00000008073405a6 in nsBaseAppShell::Run() () from /usr/local/lib/waterfox/libxul.so
#23 0x00000008083c4799 in nsAppStartup::Run() () from /usr/local/lib/waterfox/libxul.so
#24 0x000000080845b58d in XREMain::XRE_mainRun() () from /usr/local/lib/waterfox/libxul.so
#25 0x000000080845bdcd in XREMain::XRE_main(int, char**, mozilla::BootstrapConfig const&) () from /usr/local/lib/waterfox/libxul.so
#26 0x000000080845c2e4 in XRE_main(int, char**, mozilla::BootstrapConfig const&) () from /usr/local/lib/waterfox/libxul.so
#27 0x000000000102c6e9 in main ()
(gdb) c
Continuing.
[LWP 101705 of process 12385 exited]
```
…

```text
[LWP 101868 of process 12385 exited]

Program terminated with signal SIGSEGV, Segmentation fault.
The program no longer exists.
(gdb) q
$ 
```

In full: 

- [2018-11-25 08:36 terminal output.txt](https://github.com/MrAlex94/Waterfox/files/2613008/2018-11-25.08.36.terminal.output.txt)

I have outputs from two earlier runs, if required.

**Comment by grahamperrin** [comment]
I wonder, Mozilla bug [1324883 - Media cache (HTML5) not cleared when tab closed or Ctrl+F5](https://bugzilla.mozilla.org/show_bug.cgi?id=1324883)

- [fixed in Firefox 58](https://bugzilla.mozilla.org/show_bug.cgi?id=1324883#c56)
- subsequently backed out …

#### SnowWhiteKiller

https://github.com/MrAlex94/Waterfox/blob/051177fe7df0c23b54b625968aa7611bd76b2112/xpcom/base/nsCycleCollector.cpp#L8

https://github.com/MrAlex94/Waterfox/blob/051177fe7df0c23b54b625968aa7611bd76b2112/xpcom/base/nsCycleCollector.cpp#L26

https://github.com/MrAlex94/Waterfox/blob/051177fe7df0c23b54b625968aa7611bd76b2112/xpcom/base/nsCycleCollector.cpp#L2625

**Comment by jbeich** [comment]
@grahamperrin, try dropping [files/patch-bug1453127](https://github.com/freebsd/freebsd-ports/blob/319ab8ffcf01f38eb64a200d2275ebad205adc6d/www/waterfox/files/patch-bug1453127) (supposed to fix [CVE-2018-5156](https://www.mozilla.org/en-US/security/advisories/mfsa2018-15/#CVE-2018-5156)). Probably, the backport is incorrect.

**Comment by grahamperrin** [comment]
Thanks. Now also reading this alongside #654 (merged) for 1453127. 

----

Note to self: Poudriere is already busy building rust, and I had clumsily/blindly thrown in a patch for Waterfox based on https://hg.mozilla.org/releases/mozilla-release/diff/9941e68b5a53/dom/html/HTMLMediaElement.cpp but now I see, the bug linked from the changeset – [1369548 - DRM video taints the particular video element permanently](https://bugzilla.mozilla.org/show_bug.cgi?id=1369548) (RESOLVED FIXED in Firefox 59) – is irrelevant to (Firefox and) Waterfox on FreeBSD; no Widevine.

**Comment by grahamperrin** [comment]
> … try dropping [files/patch-bug1453127](https://github.com/freebsd/freebsd-ports/blob/319ab8ffcf01f38eb64a200d2275ebad205adc6d/www/waterfox/files/patch-bug1453127) …

Done, built, installed, no longer crashing :+1: 

I can live without a fix for CVE-2018-5156. 

Thanks again @jbeich 

----
